File: frontend/src/types/inventory.ts
export type InventoryPackage = {
  id: string;
  package_id: string;
  lot_code: string;
  product_id: string;
  sku: string;
  product_name: string;
  material_type: string;
  location: string;
  status: string;
  source_name: string;
  available: number;
  reserved: number;
  usable: number;
  unit: string;
  received_at: string | null;
  expiration_at: string | null;
  attention: string;
  sold_30d: number; daily_velocity: number; days_on_hand: number | null;
  unit_cost: number; retail_price: number; margin_pct: number | null;
  age_days: number | null; days_to_expiry: number | null;
};

export type InventoryResponse = {
  operation: "retail" | "production";
  grain: "packages";
  items: InventoryPackage[];
  facets: { statuses: string[]; material_types: string[]; locations: string[]; sources: string[] };
  summary: {
    package_count: number;
    available_quantity: number;
    reserved_quantity: number;
    hold_count: number;
    low_balance_count: number;
  };
};

export type ProductOption = { id: string; sku: string; name: string; item_type: string; base_unit: string };
export type InventoryReceipt = {
  product_id: string; package_id: string; lot_code: string; quantity: number; unit: string;
  location: string; source_name: string; manifest_reference: string; lab_testing_state: string;
  coa_reference: string; notes: string;
};
export type InventoryReceiptHistoryItem = {
  transaction_id: string; lot_id: string; product_name: string; package_id: string;
  quantity: number; unit: string; manifest_reference: string; source_name: string;
  actor: string; received_at: string;
};
export type InventoryAdjustment = { lot_id: string; adjustment_type: "incremental" | "set_quantity"; quantity: number; reason: string; reason_note: string };
export type AuditSummary = { id: string; audit_number: string; operation_type: "retail" | "production"; status: string; scope_label: string; blind_count: boolean; recount_tolerance: number; started_at: string; completed_at: string | null; created_by: string; scanned_count:number; total_count:number; recount_count:number };
export type AuditLine = { id: string; lot_id: string; product_name: string; package_id: string; location: string; expected_quantity: number | null; counted_quantity: number | null; variance_quantity: number; recount_required: boolean; unit: string; reason: string; notes: string; first_count_quantity: number | null; recount_quantity: number | null; counted_by: string; sku_or_upc: string; lot_code: string; metrc_package: string; primary_code: string; unit_cost: number; retail_price: number };
export type AuditScan = { id: string; raw_code: string; match_status: string; scan_stage: string; scanned_by: string; scanned_at: string };
export type AuditEvent = { id: string; action: string; actor: string; occurred_at: string; changes_json: string };
export type AuditDetail = { audit: AuditSummary; lines: AuditLine[]; scans: AuditScan[]; events: AuditEvent[] };
export type PackageLineage = { lot: { lot_id: string; lot_code: string; compliance_package_id: string; product_name: string; balance: number; unit: string }; created_by: null | { run_number: string; action_type: string; parents: { lot_id: string; lot_code: string; product_name: string; quantity: number; unit: string }[] }; used_by: { run_number: string; action_type: string; quantity_consumed: number; unit: string; outputs: { lot_id: string | null; lot_code: string; inventory_quantity: number; inventory_unit: string; purpose: string }[] }[] };
export type PlantPhase = "clone" | "seedling" | "vegetative" | "flowering" | "harvested" | "destroyed";
export type CultivationPlant = { id: string; plant_tag: string; strain_name: string; phase: PlantPhase; room_code: string; source_lot_id: string | null; mother_plant_tag: string; planted_at: string | null; estimated_harvest_date: string | null; retired_at: string | null; notes: string };
export type PlantEvent = { id: string; event_type: string; from_value: string; to_value: string; reason: string; notes: string; actor: string; occurred_at: string };

File: backend/app/schemas/inventory.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class InventoryPackage(BaseModel):
    id: str
    package_id: str
    lot_code: str
    product_id: str
    sku: str
    product_name: str
    material_type: str
    location: str
    status: str
    source_name: str = ""
    available: float
    reserved: float
    usable: float
    unit: str
    received_at: datetime | None = None
    expiration_at: datetime | None = None
    attention: str
    sold_30d: float = 0.0
    daily_velocity: float = 0.0
    days_on_hand: float | None = None
    unit_cost: float = 0.0
    retail_price: float = 0.0
    margin_pct: float | None = None
    age_days: float | None = None
    days_to_expiry: float | None = None


class InventoryFacets(BaseModel):
    statuses: list[str]
    material_types: list[str]
    locations: list[str]
    sources: list[str] = []


class InventorySummary(BaseModel):
    package_count: int
    available_quantity: float
    reserved_quantity: float
    hold_count: int
    low_balance_count: int


class InventoryResponse(BaseModel):
    operation: Literal["retail", "production"]
    grain: str = "packages"
    items: list[InventoryPackage]
    facets: InventoryFacets
    summary: InventorySummary


class ProductOption(BaseModel):
    id: str
    sku: str
    name: str
    item_type: str
    base_unit: str


class InventoryReceiptCreate(BaseModel):
    product_id: str
    lot_code: str = ""
    package_id: str = ""
    quantity: float = Field(gt=0)
    unit: str
    location: str = "RECEIVING"
    source_name: str = ""
    manifest_reference: str = ""
    lab_testing_state: str = ""
    coa_reference: str = ""
    expiration_at: datetime | None = None
    notes: str = ""


class InventoryReceiptResult(BaseModel):
    lot_id: str
    transaction_id: str
    operation: Literal["retail", "production"]
    status: str


class InventoryReceiptHistoryItem(BaseModel):
    transaction_id: str
    lot_id: str
    product_name: str
    package_id: str
    quantity: float
    unit: str
    manifest_reference: str
    source_name: str
    actor: str
    received_at: datetime


class RetailSaleImportLine(BaseModel):
    source_record_id: str
    sold_at: datetime
    quantity: float
    product_id: str | None = None
    sku: str = ""
    product_name: str
    net_sales: float = 0.0


class RetailSalesImport(BaseModel):
    source_system: str
    import_batch_id: str = ""
    lines: list[RetailSaleImportLine] = Field(min_length=1, max_length=10000)


class RetailSalesImportResult(BaseModel):
    imported: int
    skipped_duplicates: int
    unmapped_products: int


class InventoryAdjustmentCreate(BaseModel):
    lot_id: str
    package_id: str = ""
    adjustment_type: Literal["incremental", "set_quantity"]
    quantity: float
    reason: str
    reason_note: str = ""
    sync_to_metrc: bool = False
    bypass_state_system: bool = False
    reviewed: bool = False


class InventoryAdjustmentResult(BaseModel):
    transaction_id: str
    lot_id: str
    previous_quantity: float
    delta: float
    final_quantity: float
    reserved_quantity: float
    unit: str
    reason: str
    metrc_status: str = "not_configured"
    traceability_transaction_id: str = ""


class InventoryAuditCreate(BaseModel):
    audit_number: str
    scope_label: str = "Full facility"
    notes: str = ""
    lot_ids: list[str] = Field(default_factory=list)
    blind_count: bool = True
    recount_tolerance: float = Field(default=0.0, ge=0)


class InventoryAuditSummary(BaseModel):
    id: str
    audit_number: str
    operation_type: Literal["retail", "production"]
    status: str
    scope_label: str
    blind_count: bool
    recount_tolerance: float
    started_at: datetime
    completed_at: datetime | None
    created_by: str
    scanned_count: int = 0
    total_count: int = 0
    recount_count: int = 0


class InventoryAuditLineItem(BaseModel):
    id: str
    lot_id: str
    product_name: str
    package_id: str
    location: str
    expected_quantity: float | None
    counted_quantity: float | None
    variance_quantity: float
    recount_required: bool
    unit: str
    reason: str
    notes: str
    first_count_quantity: float | None = None
    recount_quantity: float | None = None
    counted_by: str = ""
    sku_or_upc: str = ""
    lot_code: str = ""
    metrc_package: str = ""
    primary_code: str = ""
    unit_cost: float = 0.0
    retail_price: float = 0.0


class InventoryAuditScanItem(BaseModel):
    id: str
    raw_code: str
    match_status: str
    scan_stage: str
    scanned_by: str
    scanned_at: datetime


class InventoryAuditEventItem(BaseModel):
    id: str
    action: str
    actor: str
    occurred_at: datetime
    changes_json: str


class InventoryAuditDetail(BaseModel):
    audit: InventoryAuditSummary
    lines: list[InventoryAuditLineItem]
    scans: list[InventoryAuditScanItem] = Field(default_factory=list)
    events: list[InventoryAuditEventItem] = Field(default_factory=list)


class InventoryAuditCount(BaseModel):
    line_id: str
    counted_quantity: float = Field(ge=0)
    reason: str = ""
    notes: str = ""


class InventoryAuditCounts(BaseModel):
    counts: list[InventoryAuditCount]


class InventoryAuditStatusChange(BaseModel):
    status: Literal["draft", "in_progress", "paused", "stopped", "cancelled"]


class InventoryAuditComplete(BaseModel):
    post_adjustments: bool = False


class InventoryAuditScanPreview(BaseModel):
    raw_code: str
    recount: bool = False


class InventoryAuditScanCount(InventoryAuditScanPreview):
    quantity: float = Field(ge=0)
    reason: str = ""
    notes: str = ""


class RetailAuditSnapshotImport(BaseModel):
    reference: str
    rows: list[dict]
    mapping: dict[str, str]


class PackageLineageLot(BaseModel):
    lot_id: str
    lot_code: str
    product_name: str = ""
    quantity: float | None = None
    unit: str = ""


class PackageLineageCreation(BaseModel):
    run_number: str
    action_type: str
    parents: list[PackageLineageLot]


class PackageLineageChildOutput(BaseModel):
    lot_id: str | None = None
    lot_code: str
    product_id: str
    inventory_quantity: float
    inventory_unit: str
    purpose: str


class PackageLineageUse(BaseModel):
    run_number: str
    action_type: str
    quantity_consumed: float
    unit: str
    outputs: list[PackageLineageChildOutput]


class PackageLineage(BaseModel):
    lot: dict
    created_by: PackageLineageCreation | None
    used_by: list[PackageLineageUse]
